skrub/_expressions/_evaluation.py

```diff
@@ -0,0 +1,711 @@
+        self.value = value
+
+
+class _FindNode(_ExprTraversal):
+    def __init__(self, predicate=None):
+        self.predicate = predicate
+
+    def handle_expr(self, e, *args, **kwargs):
+        if self.predicate is None or self.predicate(e):
+            raise _Found(e)
+        yield from super().handle_expr(e, *args, **kwargs)
+
+    def handle_choice(self, choice):
+        if self.predicate is None or self.predicate(choice):
+            raise _Found(choice)
+        yield from super().handle_choice(choice)
+
+
+def find_node(obj, predicate=None):
+    try:
+        _FindNode(predicate).run(obj)
+    except _Found as e:
+        return e.value
+    return None
+
+
+def find_X(expr):
+    return find_node(expr, lambda e: isinstance(e, Expr) and e._skrub_impl.is_X)
+
+
+def find_y(expr):
+    return find_node(expr, lambda e: isinstance(e, Expr) and e._skrub_impl.is_y)
+
+
+def find_node_by_name(expr, name):
+    def pred(obj):
+        if isinstance(obj, Expr):
+            return obj._skrub_impl.name == name
+        return getattr(obj, "name", None) == name
+
+    return find_node(expr, pred)
+
+
+def needs_eval(obj, return_node=False):
+    try:
+        node = find_node(obj)
+    except CircularReferenceError:
+        node = None
+    needs = node is not None
+    if return_node:
+        return needs, node
+    return needs
+
+
+class _FindConflicts(_ExprTraversal):
+    """Find duplicate names or if 2 nodes are marked as X or y."""
+
+    def __init__(self):
+        self._names = {}
+        self._x = {}
+        self._y = {}
+
+    def handle_expr(self, e, *args, **kwargs):
+        self._add(
+            e,
+            getattr(e._skrub_impl, "name", None),
+            e._skrub_impl.is_X,
+            e._skrub_impl.is_y,
+        )
+        yield from super().handle_expr(e, *args, **kwargs)
+
+    def handle_choice(self, choice):
+        self._add(choice, getattr(choice, "name", None), False, False)
+        yield from super().handle_choice(choice)
+
+    def _conflict_error_message(self, conflict):
+        first, second = conflict["nodes"]
+        if conflict["reason"] == "is_X":
+            return (
+                "Only one node can be marked with `mark_as_x()`. "
+                "2 different objects were marked as X:\n"
+                f"first object that used `.mark_as_x()`:\n{first!r}\n"
+                f"second object that used `.mark_as_x()`:\n{second!r}"
+            )
+        if conflict["reason"] == "is_y":
+            return (
+                "Only one node can be marked with `mark_as_y()`. "
+                "2 different objects were marked as y:\n"
+                f"first object that used `.mark_as_y()`:\n{first!r}\n"
+                f"second object that used `.mark_as_y()`:\n{second!r}"
+            )
+        if conflict["reason"] == "name":
+            name = conflict["name"]
+            return (
+                f"Choice and node names must be unique. The name {name!r} was used "
```

Comment:
Here we have a problem: as I work interactively in a vscode/jupyter, I modify a cell, and rerun it multiple times.

As a consequence, I end up defining nodes with the same name, get an error which prevents me from working:
```
ValueError: Choice and node names must be unique. The name 'encoder' was used for 2 different objects:
first object using the name 'encoder':
choose_from({'MinHash': MinHashEncoder(), 'LSA': StringEncoder()}, name='encoder')
second object using the name 'encoder':
choose_from({'MinHash': MinHashEncoder(), 'LSA': StringEncoder()}, name='encoder')
```

That's a big blocker

Reply:
but I guess that means you're running something like `table = table.skb.apply(encoder)` several times? that would always be a problem, even if it was `table = encoder.fit_transform(table)` for example? or I'm guessing the content of the cell wrong

Reply:
That was me being stupid